### PR TITLE
Fix inconsistencies between single key operations (`merge`, `set`) and multi key operations (`mergeCollection`, `multiSet`)

### PR DESCRIPTION
**Pull request by chrispader**
@paultsimura 

### Details

This PR aims to remove inconsistencies between "single key operations" (`merge`, `set`, ...) and "multi key operations" (`mergeCollection`, which uses `multiMerge` under the hood, `multiSet`, ...).

When using `mergeCollection` nested null values would not be removed from storage, because we already removed the `null` values from the changes in that are then later (natively) merged in the provider. Keeping the `null` values in the changes sent to the storage layer ensures, that the same data will be written, as when we used `merge`.

This PR also makes sure that nested `null` values that we get from cache are never sent to the user (e.g. with `withOnyx`, `useOnyx` or a callback). Cache in Onyx will still contain `null` values for performance reasons, as in #411. We can still use the original data to compare data to prevent unnecessary re-renders, but we should never provide `null` values to `withOnyx` props or callbacks.

This PR also adds some more unit tests to check for these new changes.

### Related Issues

$ https://github.com/Expensify/react-native-onyx/issues/516
$ https://github.com/Expensify/react-native-onyx/issues/514

### Automated Tests

Run Jest tests especially the ones in `oynxTest` regarding `mergeCollection`.

### Manual Tests

Somewhere in E/App or any other test project:

1. Set the value for a key with an initial object
2. Run `Onyx.mergeCollection` with nested null values for already existing keys.
3. Check the value from either the callback from `Onyx.connect` or the valuef from  `useOnyx`/`withOnyx`
4. The value should not contain the removed key anymore.

### Author Checklist

- [x] I linked the correct issue in the `### Related Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

Screenshot 2024-03-28 at 17 48 27

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

**Comment by tgolen** [comment]
I'm placing this on HOLD right now in order for QA to be done on the latest version of Onyx without more un-QAed PRs from being merged.

**Comment by chrispader** [comment]
> I'm placing this on HOLD right now in order for QA to be done on the latest version of Onyx without more un-QAed PRs from being merged.

I integrated #515 into this PR, so this should be the last PR in Onyx before we can bump it in E/App.

**Comment by chrispader** [comment]
> @chrispader I'm getting this error when launching the App:
> 
> image

@paultsimura fixed! I confused `propsToPass` in `withOnyx.js` with the Onyx keys being mapped to the component...

@tgolen @MonilBhavsar i'm gonna update the description of this PR and clean it up a bit, but this should be ready for review!

**Comment by tgolen** [comment]
We can wait for the C+ review in the Onyx bump PR I think 👍

File: lib/storage/providers/MemoryOnlyProvider.ts [hunk 1]
```diff
@@ -68,11 +68,7 @@ const provider: StorageProvider = {
     multiSet(pairs) {
         const setPromises = _.map(pairs, ([key, value]) => this.setItem(key, value));
 
-        return new Promise((resolve) => {
-            Promise.all(setPromises).then(() => {
-                resolve(undefined);
-            });
-        });
+        return Promise.all(setPromises).then(() => undefined);
```
Comment (chrispader) [inline lib/storage/providers/MemoryOnlyProvider.ts:71]: I just noticed this return could be simplified. Not related to this PR

File: lib/OnyxUtils.js [hunk 2]
```diff
@@ -518,6 +518,8 @@ function keysChanged(collectionKey, partialCollection, notifyRegularSubscibers =
         // We prepare the "cached collection" which is the entire collection + the new partial data that
         // was merged in via mergeCollection().
         const cachedCollection = getCachedCollection(collectionKey);
+        const cachedCollectionWithoutNestedNulls = utils.removeNestedNullValues(cachedCollection);
+        // const cachedCollectionWithoutNestedNulls = cachedCollection;
```
Comment (tgolen) [inline lib/OnyxUtils.js:522]: Can this be removed?

**Review by tgolen — CHANGES_REQUESTED** [review]

File: lib/OnyxUtils.js (hunk 2 above)
Comment (chrispader) [reply lib/OnyxUtils.js:522]: ah yes, thanks!

**Review by tgolen — APPROVED** [review]

**Comment by tgolen** [comment]
@MonilBhavsar Can you please review this?

**Comment by MonilBhavsar** [comment]
Yes on it, sorry for delay!

File: lib/OnyxUtils.js [hunk 3]
```diff
@@ -1086,20 +1092,23 @@ function removeNullValues(key, value) {
...
+    return _.reduce(
+        data,
+        (pairs, value, key) => {
+            const {value: valueAfterRemoving, wasRemoved} = removeNullValues(key, value, shouldRemoveNestedNulls);
 
-        keyValuePairs.push([key, valueAfterRemoving]);
-    });
+            if (!wasRemoved) {
+                pairs.push([key, valueAfterRemoving]);
+            }
```
Comment (MonilBhavsar) [inline lib/OnyxUtils.js:1106]: NAB, we can return early if wasRemoved is true.

```js
if (wasRemoved) { return pairs; }
return pairs.push([key, valueAfterRemoving]);
```

**Review by MonilBhavsar — APPROVED** [review]
Looks good!

File: lib/OnyxUtils.js (hunk 3 above)
Comment (chrispader) [reply lib/OnyxUtils.js]: updated!

**Comment by chrispader** [comment]
can we merge this yet? @tgolen @MonilBhavsar

**Comment by mountiny** [comment]
Started a thread [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1713204411332439)

**Comment by chrispader** [comment]
merged `main`!

**Comment by chrispader** [comment]
> @chrispader could you please merge `main` again? The `lib/OnyxUtils` has now migrated to TS.

done! can't thoroughly test it right now, will do in the evening. lmk if something doesn't work

**Comment by chrispader** [comment]
tested the PR, and everything works as before. @mountiny @tgolen feel free to review/merge this whenever it's off HOLD

**Comment by mountiny** [comment]
@chrispader are there no manual tests that could be done here? 

@rinej seems like the reassure action here is not happy yet, could you check out the failure?

**Comment by chrispader** [comment]
> @chrispader are there no manual tests that could be done here?

hmm, not really, since we're testing for that in the Jest tests already. You could manually run the steps to check for this behavior either in the DevTools console or somewhere in E/App. I added the steps to `Manual tests`

**Comment by mountiny** [comment]
Removed the hold, lets ignore the reassure tests as we are just setting them up in the repo to make sure the lower level functions stay performant, however the setup does not look 100% yet, @rinej from Callstack will investigate on Monday

@tgolen @MonilBhavsar please continue with the review

**Review by tgolen — DISMISSED** [review]

File: lib/OnyxUtils.ts [hunk 4]
```diff
@@ -752,7 +756,8 @@ function sendDataToConnection<TKey extends OnyxKey>(mapping: Mapping<TKey>, val:
         return;
     }
 
-    (mapping as DefaultConnectOptions<TKey>).callback?.(val, matchedKey as TKey);
+    const valWithoutNestedNulls = utils.removeNestedNullValues(val);
```
Comment (MonilBhavsar) [inline lib/OnyxUtils.ts:759]: NAB, we could use valuesWithoutNestedNulls

**Review by MonilBhavsar — APPROVED** [review]

File: lib/OnyxUtils.ts (hunk 4 above)
Comment (chrispader) [reply lib/OnyxUtils.ts]: updated!

**Review by MonilBhavsar — APPROVED** [review]
Thanks!

**Comment by chrispader** [comment]
> @tgolen I believe it's only pending your re-approval now?

@tgolen already approved, i just changed a minor NAB mentioned by @MonilBhavsar. no critical change

**Comment by tgolen** [comment]
haha, OK. Here I go!

**Comment by roryabraham** [comment]
@francoisl [reported](https://github.com/Expensify/App/pull/41056#issuecomment-2093822668) that we had to roll back Onyx due to a regression in this PR. @chrispader can you please work on fixing this issue and creating a new PR to upgrade Onyx? This is currently preventing us from deploying other Onyx changes.

**Comment by chrispader** [comment]
Looking into fixing the regression and bumping Onyx today/tomorrow!

**Comment by chrispader** [comment]
Sorry for the delay 🙈 Working on this now!

**Comment by chrispader** [comment]
Fixed in https://github.com/Expensify/App/pull/42057

**Pull request merged**